### DISCUSSION TITLE
"date" and "modified" date, listing sorted by either

### DISCUSSION THREAD

**Discussion by danmackinlay** [quarto-dev/quarto-cli#7854] [Q&A] [answered]
### Description

In the "notebook", as opposed to "blog" format, we might care about the "last modified" date *and* the "first modified
Examples of projects which use this format to a great or lesser extent:

http://bactra.org/notebooks/
https://gwern.net/
https://danmackinlay.name/notebook.html

It would be nice to support this in Quarto, by admitting an extra data ("modified-date") and/or just arbitrary date attributes for pages and for listings and RSS feeds. Is this possible?

**Comment by danmackinlay** [comment]
Related post from the Q&A section:

* [Last-modified when using github actions to publish · quarto-dev/quarto-cli · Discussion #5899](https://github.com/quarto-dev/quarto-cli/discussions/5899)
* [Add modified-date of blogpost to pdf · quarto-dev/quarto-cli · Discussion #5570](https://github.com/quarto-dev/quarto-cli/discussions/5570)
* [HTML: Is it possible to add a date with "last-modified" field in the sidebar? · quarto-dev/quarto-cli · Discussion #6671](https://github.com/quarto-dev/quarto-cli/discussions/6671)
* [Is using modified date in the RSS feed instead of first\_published date possible? · quarto-dev/quarto-cli · Discussion #4428](https://github.com/quarto-dev/quarto-cli/discussions/4428)

**Comment by danmackinlay** [comment]
Hmm, after diving into the docs, it seems that a [custom listing](https://quarto.org/docs/websites/website-listings-custom.html#sorting-filtering-and-pagination) can do this; I will write one for myself. But also I am interested in knowing if this is the kind of thing that mainline quarto wishes to do

**Reply by danmackinlay** [reply]
The real trick here, it seems to me, is not the custom "modified" metadata, which works fine, but rather displaying the modified date on the indificual articles.
I guess we use the `template` [HTML Options](https://quarto.org/docs/reference/formats/html.html#rendering) for that.

**Reply by danmackinlay** [reply]
in fact, while trying to produce a template which did this, I discovered that it is already supported; if we inspect [/src/resources/formats/html/templates/title-metadata.html](https://github.com/quarto-dev/quarto-cli/blob/31571bcfb398868308d78633f8c2ad00955650b2/src/resources/formats/html/templates/title-metadata.html) we see that the `date-modified` field does what we want; it is displayed on individual articles, and we can easily sort by it in the listings

**Reply by danmackinlay** [reply]
~~The field may be referenced from inside the listing template as `date_modified` (i.e. with dash changed into underscore)~~

Incorrect. It turns out that the correct way of accessing the data is `item['date-modified']`

**Reply by danmackinlay** [reply]
In fact the listings *do* support displaying the modified date. The default implementation is a little confusing; both the created and modified dates are displayed with no context to explain what either of them are, e.g.
<img width="607" alt="Screenshot 2023-12-24 at 15 22 37" src="https://github.com/quarto-dev/quarto-cli/assets/21740/d885fbfe-bf79-47e7-8d48-a4b3ba76e31d">
This could possibly be fixed with some custom CSS.

**Comment by mcanouil** [comment]
I apologise but this thread is really quite confusing.
It feels like everything is already in the documentation and in Quarto since before this thread was even opened.

Also, mixing various subjects and kind of switching from one to another makes it hard to understand what's the point of the thread and if you a actually found what you were looking for.
(You can edit your initial post to structure it for example using headings or anything)

- `date-modified` is a variable added in the metadata title block in 1.4 for HTML format
- Date fields supports some "magic" keywords to get particular dates such as `last-modified` or `now`.
- listings can use any fields defined for each documents or at the directory level via `_metadata.yml`
- to add fields in listing to allow to sort or search for it, there is no need to define a custom listing via a template, simply add the fields (<https://quarto.org/docs/websites/website-listings.html#customizing-fields>)

**Reply by danmackinlay** [reply]
Thanks @mcanouil !  Those are helpful answers.

as for the bit about this thread being already in the documentation...
What you are saying seems to me to be strictly correct, in that if I search for the string "date-modified" I discovered that it is listed as metadata for posts enabled by default. So this is not actually a feature request, it turns out. Apart from that misclassification, I think this thread probably corresponds to what people would actually search for when attempting to answer this question for themselves. Or, if not, how would you phrase it differently?

I suggest that the existence of the thread about sorting posts by modified date is not so confusing. Asking questions to discover stuff in informal n00b language is a pretty normal and I think productive way of community discovery of features and terminology in open-source projects. Especially when it is hard to discover organically, which I think in this case it is not.
If, as I did, you search for "date modified" those docs pages surely are somewhere in the search results, but there are a *lot* of search results the ordering is not useful That is probably why, on this discussion group there are several posts by people (including me) trying to work out how to sort by modified date; the answer is very discoverable when you know what it is, but otherwise, not so much.
I appreciate this issue might seem straightforward for the people intimate with the project but for outsiders it seems to me that there is value in having a post on the discussion board which ranks high in the search results when you use the noob keywords, which this seems to.

I'm happy to commit to putting a clearer summary of my understanding at the end when I've actually made it all work. Hold on though, it isn't working yet! This is a live issue, at least for me.

**Reply by mcanouil** [reply]
I am not saying your posts should not exist or anything like that.
Only that even for someone that knows a lot about Quarto it is hard to connect the dots as it feels like it's notes from your train of thoughts which might be not that clear to future readers.

So yes, if at some point you can summarise and possibly make suggestions on how to improve documentation that would be awesome.

**Reply by danmackinlay** [reply]
Totally get that @mcanouil , thanks! 

Follow on: are you totally sure about this part of your comment?

> to add fields in listing to allow to sort or search for it, there is no need to define a custom listing via a template, simply add the fields (https://quarto.org/docs/websites/website-listings.html#customizing-fields)

I have tested this by adding in a custom listing:

```yaml
listing:
  page-size: 50
  sort: "date-modified desc"
  type: default
  categories: true
  sort-ui: true
  filter-ui: true
  contents:
    - notebook
  feed: true
  # template: theme/custom_listing.ejs
  fields: [date, date-modified, title]
  field-types:
    date: date
    date-modified: date
  field-display-names:
    title: "Title"
    date: "Created"
    date-modified: "Modified"
```

However, unless I _uncomment_ the line which uses the custom listing template, the `date-modifed` stuff is ignored. The listing is not sorted by date modified and `Modified` does not appear in the sort-ui dropdown.

**Comment by danmackinlay** [comment, marked as answer]
To answer my own question and to summarise and clarify the above discussion

1. Quarto 1.4 does support the use of a `date-modified` metadata field, which, if used correctly will allow you to sort by date-modified, as well as date
2.  I created an [MWE demonstrating how to do this](https://github.com/danmackinlay/quarto_date_modified)
3. There is probably a bug in quarto, because contrary to the [documentation](https://quarto.org/docs/websites/website-listings.html#customizing-fields) and to @mcanouil 's [helpful comment](https://github.com/quarto-dev/quarto-cli/discussions/7854#discussioncomment-7936934), this `date-modified` stuff appears to only work if you create a custom template,  [as seen in my MWE](https://github.com/danmackinlay/quarto_date_modified)

**Reply by danmackinlay** [reply]
Corresponding bug report https://github.com/quarto-dev/quarto-cli/issues/8198

**Reply by danmackinlay** [reply]
Update: @dragonstyle points out correctly that I was missing the full sort-ui keys, i.e. `sort-ui: [date, date-modified, title]`, so this is not in fact a bug.
